src/NeoSharp.Core/Messaging/MessageHandlerProxy.cs

```diff
@@ -1,153 +1,69 @@
-        private readonly IReadOnlyDictionary<Type, Delegate> _messageHandlerInvokers;
-
         #endregion
 
-        /// <summary>
-        /// Constructor
-        /// </summary>
-        /// <param name="container">Container</param>
-        /// <param name="messageHandlerTypes">MessageHandler types</param>
-        /// <param name="logger">Logger</param>
-        public MessageHandlerProxy(IContainer container, IEnumerable<Type> messageHandlerTypes, ILogger<MessageHandlerProxy> logger)
+        #region Constructor 
+        public MessageHandlerProxy(
+            IEnumerable<IMessageHandler> messageHandlers, 
+            ILogger<MessageHandlerProxy> logger)
         {
-            _container = container;
+            _messageHandlers = messageHandlers;
             _logger = logger;
-            _messageHandlerInvokers = messageHandlerTypes
-                .Select(CreateMessageHandlerInvoker)
-                .ToDictionary(x => x.MessageType, x => x.MessageHandlerInvoker);
         }
+        #endregion
 
-        /// <summary>
-        /// Handle message
-        /// </summary>
-        /// <param name="message">Message</param>
-        /// <param name="sender">Sender</param>
-        /// <returns>Task</returns>
-        public Task Handle(Message message, IPeer sender)
+        #region IMessageHandler implementation 
+        public async Task Handle(Message message, IPeer sender)
         {
             if (message == null) throw new ArgumentNullException(nameof(message));
 
-            var messageType = message.GetType();
-            var messageHandler = ResolveMessageHandler(messageType);
-            var messageHandlerName = messageHandler.GetType().Name;
-            var startedAt = LogMessageHandlingStart(messageHandlerName);
-            var messageHandlerInvoker = GetMessageHandlerInvoker(messageType);
-            var handleMessageTask = (Task)messageHandlerInvoker.DynamicInvoke(messageHandler, message, sender);
-
-            return handleMessageTask.ContinueWith(t =>
-            {
-                if (t.IsCompletedSuccessfully)
-                {
-                    LogMessageHandlingEnd(startedAt, messageHandlerName);
-                    return;
-                }
-
-                if (t.IsFaulted)
-                {
-                    LogMessageHandlingEnd(startedAt, messageHandlerName, handleMessageTask.Exception);
-                    return;
-                }
-            });
-        }
+            var isHandled = false;
 
-        private static (Type MessageType, Delegate MessageHandlerInvoker) CreateMessageHandlerInvoker(Type messageHandlerType)
-        {
-            const string handleMethodName = nameof(IMessageHandler<Message>.Handle);
-            const BindingFlags bindingFlags = BindingFlags.FlattenHierarchy | BindingFlags.Instance | BindingFlags.Public;
+            var messageHandler = this._messageHandlers.SingleOrDefault(x => x.CanHandle(message));
```

Comment:
As I said before, this change alters the behavior we introduced before (message handlers per message vs singletons).
I am not a fan to pass IContainer as a dependency and it is the very last resort for me but, as you can see, in this case we need to have dynamic routing based on message type and to create message handlers on demand. I suggest to keep it in this was way to avoid possible side effects in message handlers. At present, any message handler can be disposed after a message is handled. My recommendation is reverting these changes (untyped message handlers, message handlers as singletons, iteration trough all message handlers for every message) until we have a full identical solution. I prefer not to replace the working code with partial one backed by the issue.

Reply:
I would like to remember that this code was not working and no one was understanding due the fact that was not fowling the standards. 
Thanks for you opinion. It is important for us to move foward but this code was not fowling the code guidelines defined by @gubanotorious where we should use code patterns. This code was weak, using Reflection with a lot of loops to create the MesssageHandler and MessageInvokers. Now it's just have one loop and you prefer the old code.

I'm sure we can improve this code in the refactor phase. For now the code is working with no bugs.